src/treelearner/serial_tree_learner.cpp

```diff
@@ -817,41 +841,61 @@ void SerialTreeLearner::Split(Tree* tree, int best_leaf, int* left_leaf, int* ri
   // init the leaves that used on next iteration
+  int depth = tree->leaf_depth(*left_leaf);
+  #ifdef DEBUG
+  CHECK(depth == tree->leaf_depth(*right_leaf));
+  #endif
   if (best_split_info.left_count < best_split_info.right_count) {
-    smaller_leaf_splits_->Init(*left_leaf, data_partition_.get(), best_split_info.left_sum_gradient, best_split_info.left_sum_hessian);
-    larger_leaf_splits_->Init(*right_leaf, data_partition_.get(), best_split_info.right_sum_gradient, best_split_info.right_sum_hessian);
+    smaller_leaf_splits_->Init(*left_leaf, data_partition_.get(),
+                               best_split_info.left_sum_gradient,
+                               best_split_info.left_sum_hessian, depth);
+    larger_leaf_splits_->Init(*right_leaf, data_partition_.get(),
+                              best_split_info.right_sum_gradient,
+                              best_split_info.right_sum_hessian, depth);
   } else {
-    smaller_leaf_splits_->Init(*right_leaf, data_partition_.get(), best_split_info.right_sum_gradient, best_split_info.right_sum_hessian);
-    larger_leaf_splits_->Init(*left_leaf, data_partition_.get(), best_split_info.left_sum_gradient, best_split_info.left_sum_hessian);
-    p_right = smaller_leaf_splits_.get();
-    p_left = larger_leaf_splits_.get();
-  }
-  p_left->SetValueConstraint(best_split_info.min_constraint, best_split_info.max_constraint);
-  p_right->SetValueConstraint(best_split_info.min_constraint, best_split_info.max_constraint);
-  if (is_numerical_split) {
-    double mid = (best_split_info.left_output + best_split_info.right_output) / 2.0f;
-    if (best_split_info.monotone_type < 0) {
-      p_left->SetValueConstraint(mid, best_split_info.max_constraint);
-      p_right->SetValueConstraint(best_split_info.min_constraint, mid);
-    } else if (best_split_info.monotone_type > 0) {
-      p_left->SetValueConstraint(best_split_info.min_constraint, mid);
-      p_right->SetValueConstraint(mid, best_split_info.max_constraint);
-    }
+    smaller_leaf_splits_->Init(*right_leaf, data_partition_.get(),
+                               best_split_info.right_sum_gradient,
+                               best_split_info.right_sum_hessian, depth);
+    larger_leaf_splits_->Init(*left_leaf, data_partition_.get(),
+                              best_split_info.left_sum_gradient,
+                              best_split_info.left_sum_hessian, depth);
+  }
+
+  // when the monotone precise mode is disabled it is very easy to compute the constraints of
+  // the children of a leaf, but when it is enabled, one needs to go through the tree to do so,
+  // and it is done directly before computing best splits
+  if (!config_->monotone_precise_mode) {
+    LeafConstraints::SetChildrenConstraintsFastMethod(
+        constraints_per_leaf_, right_leaf, left_leaf,
+        best_split_info.monotone_type, best_split_info.right_output,
+        best_split_info.left_output, is_numerical_split);
+  }
```

Comment:
will this be called when mc is not set?

Reply:
Yes, this is still the same logic as before the PR was made. When there are no monotone constraints, then the max constraint for a leaf is +inf and the min constraints is -inf.